crypto/rsa/rsa_backend.c

```diff
@@ -64,22 +64,109 @@ static int collect_numbers(STACK_OF(BIGNUM) *numbers,
 int ossl_rsa_fromdata(RSA *rsa, const OSSL_PARAM params[], int include_private)
 {
     const OSSL_PARAM *param_n, *param_e,  *param_d = NULL;
-    BIGNUM *n = NULL, *e = NULL, *d = NULL;
+    const OSSL_PARAM *param_p, *param_q = NULL;
+    const OSSL_PARAM *param_derive = NULL;
+    BIGNUM *p = NULL, *q = NULL, *n = NULL, *e = NULL, *d = NULL;
     STACK_OF(BIGNUM) *factors = NULL, *exps = NULL, *coeffs = NULL;
     int is_private = 0;
+    int derive_from_pq = 0;
+    BN_CTX *ctx = NULL;
 
     if (rsa == NULL)
         return 0;
 
+    ctx = BN_CTX_new_ex(rsa->libctx);
+    if (ctx == NULL)
+        return 0;
+
+    param_derive = OSSL_PARAM_locate_const(params,
+                                           OSSL_PKEY_PARAM_RSA_DERIVE_FROM_PQ);
+    if ((param_derive != NULL)
+        && !OSSL_PARAM_get_int(param_derive, &derive_from_pq))
+        goto err;
+
     param_n = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_N);
     param_e = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_E);
     if (include_private)
         param_d = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_D);
 
-    if ((param_n != NULL && !OSSL_PARAM_get_BN(param_n, &n))
-        || (param_e != NULL && !OSSL_PARAM_get_BN(param_e, &e))
-        || (param_d != NULL && !OSSL_PARAM_get_BN(param_d, &d)))
-        goto err;
+    if (derive_from_pq) {
+        /* we need at minimum p, q and e */
+        param_p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_FACTOR1);
+        param_q = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_FACTOR2);
```

Comment:
As this codepath does not support multiprime RSA keys you should check there is no OSSL_PKEY_PARAM_RSA_FACTOR3 param here.

Reply:
what would the error be in that case?  Someone requested an RSA key, and provided the minimum needed parameters to do so.  Having extra....is odd, I grant you, but not expressly an error, as we can still create the key

Reply:
regarding the code movement below.  I don't think we can do that as the crt generation call expects the factors to have already been set (if I'm reading your comment correctly)

Reply:
A design decision was to ignore extra parameters if they are supplied.  I don't think this is an error because of that.  If we had a way to producing warnings, it could be that.

Reply:
Please no! It is totally wrong to compute incorrect values just because we are missing multiprime support here. The design decision was (at least partially) wrong and it will bite us.

The error is clear here. Someone is thinking that computing n for multiprime RSA on import is supported but you decide to ignore any further factors and compute incorrect n value and/or d value, etc.

I am more and more convinced it was wrong to expand the original scope from computing the missing CRT parameters to also compute n and d.

Reply:
It seems to me like this is probably an OTc decision